File: src-python/planarvibe/layouts/reweight.py
"""Reweight layout (iterative barycentric reweighting for equal face areas).

Literal port of static/js/layout-reweight.js.
"""

from __future__ import annotations

import math
from typing import Sequence

from .. import geometry as geo
from .. import graph as gu
from .. import metrics as metric_mod
from .. import planar_graph as pg
from .. import preprocessing
from . import tutte

REWEIGHT_CONFIG = {
    "maxOuterIters": 8,
    "pressureStep": 0.16,
    "pressureClamp": 1.20,
    "pressureBeta": 0.18,
    "pressureDeltaClamp": 0.75,
    "scaleMin": 0.25,
    "scaleMax": 10.0,
    "pressureScaleMin": 1.0,
    "pressureScaleMax": 1.25,
    "minItersBeforeStop": 8,
    "stableIterLimit": 4,
}


def _solve_weighted(state: dict, weights: dict) -> dict:
    return tutte.compute_barycentric_positions(
        state["augmented"]["graph"],
        state["outerFace"],
        {
            "adjacency": state["adj"],
            "weights": weights,
            "initOptions": tutte.default_outer_placement_options({
                "fixedOuterPos": state["fixedOuterPos"],
            }),
        },
    )


def _build_edge_to_face_map(faces: list[list[str]]) -> dict[str, list[int]]:
    out: dict[str, list[int]] = {}
    for i, face in enumerate(faces):
        n = len(face)
        for j in range(n):
            u = str(face[j])
            v = str(face[(j + 1) % n])
            k = gu.edge_key(u, v)
            out.setdefault(k, []).append(i)
    return out


def _update_face_pressures(state: dict, face_areas: list[float], face_pressure: list[float]) -> list[float]:
    opts = state["opts"]
    bounded_face_idx = state["boundedFaceIdx"]
    nxt = list(face_pressure)
    ssum = 0.0
    cnt = 0
    for fi in bounded_face_idx:
        area = face_areas[fi]
        if not math.isfinite(area) or not (area > 1e-12):
            continue
        delta = math.log(max(state["desired"], 1e-12) / max(area, 1e-12))
        delta = max(-opts["pressureDeltaClamp"], min(opts["pressureDeltaClamp"], delta))
        p = nxt[fi] + opts["pressureStep"] * delta
        p = max(-opts["pressureClamp"], min(opts["pressureClamp"], p))
        nxt[fi] = p
        ssum += p
        cnt += 1
    mean = (ssum / cnt) if cnt > 0 else 0.0
    if cnt > 0 and abs(mean) > 1e-12:
        for fi in bounded_face_idx:
            nxt[fi] -= mean
    return nxt


def _adjust_weights(state: dict, face_areas: list[float], old_weights: dict, face_pressure: list[float]) -> dict:
    edge_pairs = state["augmented"]["graph"].edge_pairs
    outer_set = {str(x) for x in state["outerFace"]}
    opts = state["opts"]
    new_weights: dict[str, float] = {}
    sum_w = 0.0
    cnt = 0

    for (u_raw, v_raw) in edge_pairs:
        u, v = str(u_raw), str(v_raw)
        k = gu.edge_key(u, v)
        w_old = old_weights.get(k)
        if not isinstance(w_old, (int, float)) or not math.isfinite(w_old) or w_old <= 0:
            w_old = 1.0

        if u in outer_set and v in outer_set:
            new_weights[k] = w_old
            continue

        faces_idx = state["e2f"].get(k, [])
        area_sum = 0.0
        area_cnt = 0
        for fi in faces_idx:
            a = face_areas[fi]
            if math.isfinite(a) and a > 0:
                area_sum += a
                area_cnt += 1
        if area_cnt == 0:
            new_weights[k] = w_old
            sum_w += new_weights[k]
            cnt += 1
            continue

        penalty = (area_sum / area_cnt) / max(state["desired"], 1e-12)
        scale = math.sqrt(penalty) if penalty > 1 else penalty
        scale = max(opts["scaleMin"], min(opts["scaleMax"], scale))

        p_sum = 0.0
        p_cnt = 0
        for fi in faces_idx:
            if fi not in state["boundedSet"]:
                continue
            p = face_pressure[fi]
            if math.isfinite(p):
                p_sum += p
                p_cnt += 1
        if p_cnt > 0 and opts["pressureBeta"] > 0:
            pressure_scale = math.exp(-opts["pressureBeta"] * (p_sum / p_cnt))
            pressure_scale = max(opts["pressureScaleMin"], min(opts["pressureScaleMax"], pressure_scale))
            scale *= pressure_scale

        w_new = w_old * scale
        w_new = max(1e-4, min(1e4, w_new))
        new_weights[k] = w_new
        sum_w += w_new
        cnt += 1

    avg = (sum_w / cnt) if cnt > 0 else 1.0
    if not (avg > 0):
        avg = 1.0
    for (u_raw, v_raw) in edge_pairs:
        ek = gu.edge_key(u_raw, v_raw)
        new_weights[ek] = (new_weights.get(ek, 1.0)) / avg

    return new_weights


def _compute_face_area_iteration_stats(face_areas: list[float], bounded_face_idx: list[int]) -> dict | None:
    values: list[float] = []
    for fi in bounded_face_idx:
        area = face_areas[fi]
        if math.isfinite(area) and area > 1e-12:
            values.append(area)
    if not values:
        return None
    total = sum(values)
    if not (total > 0):
        return None
    values = [v / total for v in values]
    values.sort()
    ideal = 1 / len(values)
    score = metric_mod.compute_distribution_quality(values)
    return {
        "score": score if (score is not None and math.isfinite(score)) else None,
        "minRatio": values[0] / ideal,
        "maxRatio": values[-1] / ideal,
        "faceCount": len(values),
    }


def _build_reweight_settings(options: dict | None) -> dict:
    raw = options or {}
    return {
        "augmentationMethod": raw.get("augmentationMethod"),
        "currentPositions": raw.get("currentPositions"),
        "onIteration": raw.get("onIteration") if callable(raw.get("onIteration")) else None,
        **REWEIGHT_CONFIG,
    }


def _build_reweight_state_from_prepared(context: dict, options: dict) -> dict:
    settings = _build_reweight_settings(options)
    if not context or not context.get("ok"):
        return gu.build_layout_error(context or {"message": "Reweight setup failed"})

    g = context["graph"]
    outer = context["augmentedOuterFace"]
    augmented = context["augmented"]
    emb_aug = augmented["embedding"]
    faces = emb_aug.get("faces") or []
    outer_face_idx = pg.find_outer_face_index(faces, outer)
    if outer_face_idx < 0 or not outer or len(outer) < 3:
        return gu.build_layout_error({
            "message": "Shared initialization produced an invalid augmented outer face",
            "graph": g,
            "augmented": augmented,
        })
    bounded_face_idx = [i for i in range(len(faces)) if i != outer_face_idx]
    if not bounded_face_idx:
        return gu.build_layout_error({
            "message": "No bounded faces", "graph": g, "outerFace": outer, "augmented": augmented,
        })

    e2f = _build_edge_to_face_map(faces)
    weights: dict[str, float] = {}
    for (u, v) in augmented["graph"].edge_pairs:
        weights[gu.edge_key(u, v)] = 1.0
    face_pressure = [0.0] * len(faces)
    bounded_set = set(bounded_face_idx)
    desired = 1 / len(bounded_face_idx)
    fixed_outer_pos = geo.filter_position_map(context["posById"], outer)
    current_pos = context["posById"]
    movable_vertices = gu.collect_movable_vertices(augmented["graph"].node_ids, outer)
    movement_scale = geo.compute_drawing_diameter(augmented["graph"].node_ids, current_pos)
    movement_tracker = gu.create_movement_convergence_tracker({
        "minItersBeforeStop": settings["minItersBeforeStop"],
        "stableIterLimit": settings["stableIterLimit"],
        "maxMoveTol": 1e-4 * movement_scale,
        "avgMoveTol": 2e-5 * movement_scale,
    })

    return {
        "ok": True,
        "opts": settings,
        "graph": g,
        "outerFace": outer,
        "augmented": augmented,
        "faces": faces,
        "boundedFaceIdx": bounded_face_idx,
        "adj": augmented["graph"].adjacency,
        "e2f": e2f,
        "weights": weights,
        "facePressure": face_pressure,
        "boundedSet": bounded_set,
        "desired": desired,
        "fixedOuterPos": fixed_outer_pos,
        "currentPos": current_pos,
        "initIters": context.get("initResult", {}).get("iters", 0) or 0,
        "movableVertices": movable_vertices,
        "movementTracker": movement_tracker,
    }


def _run_reweight_iterations(state: dict, options: dict) -> dict:
    outer = state["outerFace"]
    augmented = state["augmented"]
    faces = state["faces"]
    bounded_face_idx = state["boundedFaceIdx"]
    weights = state["weights"]
    face_pressure = state["facePressure"]
    current_pos = state["currentPos"]
    total_inner_iters = state.get("initIters") or 0
    stop_reason = "max-iters"
    performed_outer_iters = 0
    final_iteration_stats = None

    for it in range(options["maxOuterIters"]):
        prev_pos = current_pos
        inner = _solve_weighted(state, weights)
        total_inner_iters += (inner.get("iters") or 0)
        performed_outer_iters = it + 1

        pos = inner["positions"]
        current_pos = pos
        move_stats = gu.compute_position_move_stats(state["movableVertices"], prev_pos, pos, {"moveTol": 1e-9})
        movement_status = state["movementTracker"].update(
            {"maxMove": move_stats.max_move, "avgMove": move_stats.avg_move},
            it + 1,
        )

        outer_area = geo.polygon_area_abs(outer, pos)
        if not (outer_area > 1e-12):
            outer_area = 1.0
        face_areas: list[float] = [geo.polygon_area_abs(f, pos) / outer_area for f in faces]
        iter_stats = _compute_face_area_iteration_stats(face_areas, bounded_face_idx)
        final_iteration_stats = iter_stats

        face_pressure = _update_face_pressures(state, face_areas, face_pressure)
        weights = _adjust_weights(state, face_areas, weights, face_pressure)

        if movement_status.converged:
            stop_reason = movement_status.reason or "movement-converged"
            break

    final_layout = _solve_weighted(state, weights)
    total_inner_iters += (final_layout.get("iters") or 0)
    final_positions = geo.filter_position_map(final_layout["positions"], state["graph"].node_ids)

    return gu.build_layout_result({
        "graph": state["graph"],
        "outerFace": outer,
        "augmented": augmented,
        "positions": final_positions,
        "debugPositions": final_layout["positions"],
        "iters": total_inner_iters,
        "stopReason": stop_reason,
        "outerSteps": performed_outer_iters,
        "faceAreaScore": final_iteration_stats["score"] if final_iteration_stats else None,
        "boundedFaceCount": final_iteration_stats["faceCount"] if final_iteration_stats else len(bounded_face_idx),
        "faceAreaMinRatio": final_iteration_stats["minRatio"] if final_iteration_stats else None,
        "faceAreaMaxRatio": final_iteration_stats["maxRatio"] if final_iteration_stats else None,
    })


def apply_layout(graph, initial_positions: dict | None = None, options: dict | None = None) -> dict:
    settings = _build_reweight_settings(options)
    prepared = preprocessing.prepare_graph_and_layout_data(graph, {
        "failureLabel": "Reweight",
        "augmentationMethod": settings["augmentationMethod"],
        "currentPositions": initial_positions or {},
    })
    if not prepared or not prepared.get("ok"):
        return gu.build_layout_error(prepared or {"message": "Reweight failed"})
    state = _build_reweight_state_from_prepared(prepared, options)
    if not state.get("ok"):
        return state
    result = _run_reweight_iterations(state, state["opts"])
    if result.get("ok"):
        result["message"] = gu.build_layout_status_message("Reweight", {
            "outerFaceVertexCount": len(result["outerFace"]),
            "dummyCount": result["augmented"].get("dummyCount"),
            "iters": result.get("iters"),
            "outerSteps": result.get("outerSteps"),
            "stopReason": result.get("stopReason"),
        })
    return result

File: src-cpp/src/reweight.cpp
#include "layouts/reweight.hpp"
#include "layouts/tutte.hpp"

#include "geometry.hpp"
#include "graph_helpers.hpp"
#include "preprocessing.hpp"

#include <algorithm>
#include <cmath>
#include <unordered_map>
#include <unordered_set>

namespace planarvibe::layouts {

namespace {

struct RConfig {
    int max_outer_iters = 8;
    double pressure_step = 0.16;
    double pressure_clamp = 1.20;
    double pressure_beta = 0.18;
    double pressure_delta_clamp = 0.75;
    double scale_min = 0.25;
    double scale_max = 10.0;
    double pressure_scale_min = 1.0;
    double pressure_scale_max = 1.25;
    int min_iters_before_stop = 8;
    int stable_iter_limit = 4;
};

std::string ekey(const std::string& a, const std::string& b) {
    return (a < b) ? (a + "::" + b) : (b + "::" + a);
}

double poly_area_abs(const std::vector<std::string>& f, const pg::PosByStr& pos) {
    if (f.size() < 3) return 0.0;
    double total = 0.0;
    int n = (int)f.size();
    for (int i = 0; i < n; ++i) {
        auto ia = pos.find(f[i]);
        auto ib = pos.find(f[(i + 1) % n]);
        if (ia == pos.end() || ib == pos.end()) return 0.0;
        total += ia->second[0] * ib->second[1] - ib->second[0] * ia->second[1];
    }
    return std::abs(total) / 2.0;
}

} // namespace

LayoutResult reweight(const Graph& g, const pg::PosByStr* initial_positions) {
    LayoutResult r;
    RConfig C;
    preprocessing::PrepareConfig cfg;
    cfg.failure_label = "Reweight";
    cfg.current_positions = initial_positions;
    auto prep = preprocessing::prepare_graph_data(g, cfg);
    if (!prep.ok) {
        r.ok = false;
        r.message = prep.message.empty() ? "Reweight failed" : prep.message;
        return r;
    }

    // Build adjacency.
    std::unordered_map<std::string, std::vector<std::string>> adjacency;
    for (const auto& nid : prep.augmented_node_ids) adjacency[nid] = {};
    for (const auto& [u, v] : prep.augmented_edge_pairs) {
        adjacency[u].push_back(v);
        adjacency[v].push_back(u);
    }

    // Initialize via Tutte-barycentric with standard weights to get starting positions.
    auto init_weights = build_tutte_weights(prep.edge_pairs, prep.augmented_edge_pairs, prep.outer_dummy_ids);
    auto init_br = compute_barycentric_positions(prep.augmented_node_ids, adjacency,
                                                  prep.augmented_outer_face, init_weights);
    if (!init_br.ok) {
        r.ok = false;
        r.message = init_br.message.empty() ? "Reweight initialization failed" : init_br.message;
        return r;
    }
    pg::PosByStr current_pos = init_br.positions;
    pg::PosByStr fixed_outer_pos;
    for (const auto& nid : prep.augmented_outer_face) {
        auto it = current_pos.find(nid);
        if (it != current_pos.end()) fixed_outer_pos[nid] = it->second;
    }

    // Face indexing.
    const auto& faces = prep.augmented_embedding.faces;
    int outer_idx = pg::find_outer_face_index(faces, prep.augmented_outer_face);
    std::vector<int> bounded_face_idx;
    for (int i = 0; i < (int)faces.size(); ++i) if (i != outer_idx) bounded_face_idx.push_back(i);
    if (bounded_face_idx.empty()) {
        r.ok = false;
        r.message = "No bounded faces";
        return r;
    }
    std::unordered_set<int> bounded_set(bounded_face_idx.begin(), bounded_face_idx.end());
    double desired = 1.0 / bounded_face_idx.size();

    // e2f: edge_key -> list of face indices
    std::unordered_map<std::string, std::vector<int>> e2f;
    for (int i = 0; i < (int)faces.size(); ++i) {
        const auto& f = faces[i];
        int n = (int)f.size();
        for (int j = 0; j < n; ++j) {
            std::string k = ekey(f[j], f[(j + 1) % n]);
            e2f[k].push_back(i);
        }
    }

    // Weights init to 1.0 per augmented edge.
    std::unordered_map<std::string, double> weights;
    for (const auto& [u, v] : prep.augmented_edge_pairs) weights[ekey(u, v)] = 1.0;

    std::vector<double> face_pressure(faces.size(), 0.0);

    // Movable vertices for convergence tracking.
    auto movable = gh::collect_movable_vertex_names(prep.augmented_node_ids, prep.augmented_outer_face);
    double movement_scale = geo::compute_drawing_diameter(
        (int)prep.augmented_node_ids.size(),
        [&]() {
            // Build a temporary PositionMap keyed by a local index.
            PositionMap pm;
            pm.resize((int)prep.augmented_node_ids.size());
            for (int i = 0; i < (int)prep.augmented_node_ids.size(); ++i) {
                auto it = current_pos.find(prep.augmented_node_ids[i]);
                if (it != current_pos.end() && std::isfinite(it->second[0]) && std::isfinite(it->second[1])) {
                    pm.put(i, it->second[0], it->second[1]);
                }
            }
            return pm;
        }());
    gh::MovementTrackerConfig tcfg;
    tcfg.min_iters_before_stop = C.min_iters_before_stop;
    tcfg.stable_iter_limit = C.stable_iter_limit;
    tcfg.max_move_tol = 1e-4 * movement_scale;
    tcfg.avg_move_tol = 2e-5 * movement_scale;
    gh::MovementTracker tracker(tcfg);

    std::unordered_set<std::string> outer_set(prep.augmented_outer_face.begin(), prep.augmented_outer_face.end());
    int total_inner_iters = init_br.iters;

    TutteOuterPlacement placement;
    placement.fixed_outer_pos = &fixed_outer_pos;

    for (int it = 0; it < C.max_outer_iters; ++it) {
        pg::PosByStr prev_pos = current_pos;
        auto inner = compute_barycentric_positions(prep.augmented_node_ids, adjacency,
                                                    prep.augmented_outer_face, weights, placement);
        if (!inner.ok) {
            r.ok = false;
            r.message = inner.message.empty() ? "Reweight inner solve failed" : inner.message;
            return r;
        }
        total_inner_iters += inner.iters;
        current_pos = inner.positions;

        auto mv = gh::compute_position_move_stats(movable, prev_pos, current_pos);
        auto ts = tracker.update(mv, it + 1);

        double outer_area = poly_area_abs(prep.augmented_outer_face, current_pos);
        if (!(outer_area > 1e-12)) outer_area = 1.0;
        std::vector<double> face_areas(faces.size(), 0.0);
        for (int i = 0; i < (int)faces.size(); ++i) {
            face_areas[i] = poly_area_abs(faces[i], current_pos) / outer_area;
        }

        // Update face pressure.
        {
            std::vector<double> nxt = face_pressure;
            double ssum = 0.0;
            int cnt = 0;
            for (int fi : bounded_face_idx) {
                double a = face_areas[fi];
                if (!std::isfinite(a) || !(a > 1e-12)) continue;
                double delta = std::log(std::max(desired, 1e-12) / std::max(a, 1e-12));
                delta = std::max(-C.pressure_delta_clamp, std::min(C.pressure_delta_clamp, delta));
                double p = nxt[fi] + C.pressure_step * delta;
                p = std::max(-C.pressure_clamp, std::min(C.pressure_clamp, p));
                nxt[fi] = p;
                ssum += p;
                ++cnt;
            }
            double mean = (cnt > 0) ? (ssum / cnt) : 0.0;
            if (cnt > 0 && std::abs(mean) > 1e-12) {
                for (int fi : bounded_face_idx) nxt[fi] -= mean;
            }
            face_pressure = std::move(nxt);
        }

        // Adjust weights.
        {
            std::unordered_map<std::string, double> new_weights;
            double sum_w = 0.0;
            int cnt = 0;
            for (const auto& [u, v] : prep.augmented_edge_pairs) {
                std::string k = ekey(u, v);
                auto wit = weights.find(k);
                double w_old = (wit != weights.end() && std::isfinite(wit->second) && wit->second > 0) ? wit->second : 1.0;
                if (outer_set.count(u) && outer_set.count(v)) {
                    new_weights[k] = w_old;
                    continue;
                }
                const auto& fs = e2f[k];
                double area_sum = 0.0;
                int area_cnt = 0;
                for (int fi : fs) {
                    double a = face_areas[fi];
                    if (std::isfinite(a) && a > 0) { area_sum += a; ++area_cnt; }
                }
                if (area_cnt == 0) {
                    new_weights[k] = w_old;
                    sum_w += w_old;
                    ++cnt;
                    continue;
                }
                double penalty = (area_sum / area_cnt) / std::max(desired, 1e-12);
                double scale = penalty > 1 ? std::sqrt(penalty) : penalty;
                scale = std::max(C.scale_min, std::min(C.scale_max, scale));
                double p_sum = 0.0;
                int p_cnt = 0;
                for (int fi : fs) {
                    if (!bounded_set.count(fi)) continue;
                    double p = face_pressure[fi];
                    if (std::isfinite(p)) { p_sum += p; ++p_cnt; }
                }
                if (p_cnt > 0 && C.pressure_beta > 0) {
                    double pressure_scale = std::exp(-C.pressure_beta * (p_sum / p_cnt));
                    pressure_scale = std::max(C.pressure_scale_min, std::min(C.pressure_scale_max, pressure_scale));
                    scale *= pressure_scale;
                }
                double w_new = w_old * scale;
                w_new = std::max(1e-4, std::min(1e4, w_new));
                new_weights[k] = w_new;
                sum_w += w_new;
                ++cnt;
            }
            double avg = (cnt > 0) ? (sum_w / cnt) : 1.0;
            if (!(avg > 0)) avg = 1.0;
            for (const auto& [u, v] : prep.augmented_edge_pairs) {
                std::string k = ekey(u, v);
                auto wit = new_weights.find(k);
                double val = (wit != new_weights.end()) ? wit->second : 1.0;
                new_weights[k] = val / avg;
            }
            weights = std::move(new_weights);
        }

        if (ts.converged) break;
    }

    // Final solve.
    auto final_br = compute_barycentric_positions(prep.augmented_node_ids, adjacency,
                                                   prep.augmented_outer_face, weights, placement);
    if (!final_br.ok) {
        r.ok = false;
        r.message = final_br.message.empty() ? "Reweight final solve failed" : final_br.message;
        return r;
    }
    total_inner_iters += final_br.iters;

    r.positions.resize(g.n);
    for (int i = 0; i < g.n; ++i) {
        auto it = final_br.positions.find(g.node_names[i]);
        if (it != final_br.positions.end() && std::isfinite(it->second[0]) && std::isfinite(it->second[1])) {
            r.positions.put(i, it->second[0], it->second[1]);
        }
    }
    if (geo::has_position_crossings(r.positions, g.edges)) {
        r.ok = false;
        r.message = "Reweight produced a non-plane drawing";
        return r;
    }
    r.ok = true;
    r.iters = total_inner_iters;
    r.message = "Reweight layout";
    return r;
}

} // namespace planarvibe::layouts
